### PR TITLE
Add user-managed certificate support and extra destinations

### LINKED ISSUE #22: Support user managed certificates in addition to automatically managed .NET style dev certs
The current feature set is focused on .NET and Aspire compatible development certificates, but a user may have reason to use a separate certificate that they've generated and trusted manually. The extensions should support syncing arbitrary certificates into the container either by file reference or retrieving a specific certificate by thumbprint from the local certificate store where applicable (possibly Windows and Mac only).

As part of this, a user should be able to disable automatic generation of a .NET/Aspire compatible certificate and the extension should be able to sync multiple certificates. Additional destinations for the certificate and key material in the container should be configurable as part of the dev container feature.

The goal is to make this a more generic solution to dev container certificate management while still supporting an automatic flow for users that can take advantage of it.

### PR DESCRIPTION
**Pull request by dnegstad**
## Summary

Closes #22. Extends the dev-certs flow beyond the single auto-generated .NET cert: users can now sync arbitrary host-side certificates into dev containers and have cert artifacts mirrored into additional directories for non-.NET workloads (nginx, Java keystores, etc.). The default flow (single auto-generated dev cert into the .NET store + OpenSSL trust dir) is unchanged and byte-identical.

Filed #26 as a follow-up to add ECDSA / Ed25519 support — `node-forge` is RSA-only, so the v1 user-cert flow only handles RSA keys.

### Devcontainer feature options

| Option | Default | Notes |
|---|---|---|
| `trustNss` | `false` | unchanged |
| `sslCertDirs` | distro list | unchanged |
| `generateDotNetCert` | `true` | per-container opt-out of auto-generation |
| `syncUserCertificates` | `true` | per-container opt-out of host-side `userCertificates` |
| `extraCertDestinations` | `""` | comma-separated `<abs-dir>[=<format>]` entries |

Either side (feature option or VS Code host setting) can independently veto dotnet cert generation — they're ANDed.

### VS Code host settings

```jsonc
"devcontainerDevCerts.generateDotNetCert": true,
"devcontainerDevCerts.userCertificates": [
  { "name": "corp-ca",  "pemCertPath": "/abs/corp-ca.pem" },
  { "name": "staging",  "pfxPath": "/abs/staging.pfx", "pfxPassword": "..." }
]
```

`name` is constrained to `^(?!\.)[A-Za-z0-9._-]{1,64}$` both at the JSON Schema level (red squiggle in the settings editor) and at runtime via `assertValidCertName` on both sides of the IPC. CA-only entries (no key) are supported — the cert is planted in the trust dir but no PFX is synced. User certs are never added to the host OS trust store.

## Architecture

- **Cross-host IPC.** New `devcontainer-dev-certs.getAllCertMaterial({ includeDotNetDev, includeUserCerts })` returns a `CertBundle`. Legacy `getCertMaterial(autoProvision)` is preserved and now honors the new host setting; the workspace extension prefers v2 and falls back to v1 on "command not found" so an older pinned UI extension keeps working.
- **Stable filename contract for extra destinations.** Auto-generated dev cert is always written as `aspnetcore-dev.*`; user certs as `<userCertificates[].name>.*`. Thumbprint-keyed filenames are reserved for the canonical .NET store paths Kestrel needs.
- **Multi-cert install path.** `installDotNetDevCert` / `installUserCert` split the canonical write logic; `writeExtraDestination` handles `pem | key | pem-bundle | pfx | all`. After the batch, `c_rehash` runs once per unique destination directory.
- **Setup script fallback.** `setup-cert.sh` learns a `--bundle-json` form for non-VSCode scenarios that need multi-cert installs; the legacy positional form is unchanged.

## Iteration history (notable review fixups in this branch)

- Simplified `extraCertDestinations` to directory-only (no `${name}` template, no single-file mode); the cert names come from a separate config and have no reliable mapping to a per-container filename the feature could anticipate.
- Activation toast now fires only when at least one cert was actually installed this run (tracks `newInstalls` / `alreadyInstalled` / `failures` separately).
- Removed a dead `GET_CERT_COMMAND` not-found check from the v2 catch block.
- All values appended to `/etc/environment` go through a quote-and-escape helper so values containing spaces or quotes (e.g. `extraCertDestinations` with `, `-separated entries) survive `pam_env` parsing.
- `setup-cert.sh` `seq` loops are guarded against empty arrays so non-GNU `seq` builds don't error on `seq 0 -1` under `set -e`.
- Shellcheck-clean on both `install.sh` and `setup-cert.sh`.

## Test coverage

- **UI extension** (10 files, 103 tests + 5 skipped): new tests for the cert loader (round-trip + CA-only + expired flag), `exportLoadedCert`, multi-cert bundle assembly, expiry-warning emission, and the `isValidCertName` allowlist (path traversal, dotfiles, oversize, emojis).
- **Workspace extension** (2 files, 13 tests): new tests for the destinations parser edge cases and the `writeExtraDestination` format matrix including CA-only skipping, traversal rejection, and trailing-slash tolerance.
- Existing exporter / manager / generator / linuxStore / nssTrust tests pass unchanged.

## Backward compatibility

- Older workspace extension + new UI extension: legacy `getCertMaterial` still returns the original `CertMaterial` shape.
- New workspace extension + older UI extension: v2 call throws not-found; legacy fallback runs and prompts to install if the UI extension itself is missing.
- Existing `setup-cert.sh <pfx> <pem> <thumbprint>` positional invocation is preserved verbatim.

## Verification

`npm run build && npm run test` clean across both extensions. Shellcheck clean on the feature scripts. `setup-cert.sh --bundle-json` smoke-tested end-to-end with empty and populated bundles.

## Deferred (issue #26)

Forge is RSA-only — `forge.pki.certificateFromPem` rejects EC certs and `privateKeyFromPem` rejects EC keys. v1 user-cert flow accepts RSA only; ECDSA/Ed25519 will need a swap to Node `crypto` + `@peculiar/x509` + `pkijs` and is tracked separately.

https://claude.ai/code/session_01ML5UqawuoyDS9ectDcwEnC

### CODE REVIEW

File: src/vscode-workspace-extension/src/util/destinations.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    .filter((e) => e.length > 0);
+
+  for (const entry of entries) {
+    const eqIdx = entry.indexOf("=");
+    let pathPart = entry;
+    let formatPart = "all";
+    if (eqIdx >= 0) {
+      pathPart = entry.slice(0, eqIdx).trim();
+      formatPart = entry.slice(eqIdx + 1).trim() || "all";
+    }
```
[the remark this answers is not in the document]
Comment (dnegstad) [reply src/vscode-workspace-extension/src/util/destinations.ts:51]: Switched to only supporting the directory output version; single file outputs weren't intended to be supported in the first place.

File: src/devcontainer-feature/src/devcontainer-dev-certs/scripts/setup-cert.sh [hunk 2]
```diff
@@ -2,22 +2,260 @@
...
+            emit "key" "${pem_key_path}"
+            emit "pfx" "${pfx_path}"
+            if [ "${is_dir}" = "true" ]; then
+                local out="${target_dir}/${name}-bundle.pem"
+                cat "${pem_path}" > "${out}"
+                if [ -n "${pem_key_path}" ] && [ -f "${pem_key_path}" ]; then
+                    cat "${pem_key_path}" >> "${out}"
+                fi
+            fi
+            ;;
```
[the remark this answers is not in the document]
Comment (dnegstad) [reply src/devcontainer-feature/src/devcontainer-dev-certs/scripts/setup-cert.sh:152]: I've removed the individual certificate file output support as that was never an intended part of the feature.

File: src/vscode-workspace-extension/src/certInstaller.ts [hunk 3]
```diff
@@ -6,81 +6,288 @@ import {
...
+    const bundleSuffix = isBundle ? "-bundle" : "";
+    if (dest.kind === "directory") {
+      const dir = dest.path.replace(/\/$/, "");
+      return path.join(dir, `${material.name}${bundleSuffix}.${ext}`);
+    }
+    if (dest.kind === "file-template") {
+      return dest.path.replace(/\$\{name\}/g, material.name);
+    }
+    return dest.path;
+  }
```
[the remark this answers is not in the document]
Comment (dnegstad) [reply src/vscode-workspace-extension/src/certInstaller.ts]: Removed support for individual file certificate output as it wasn't an intended part of this feature.

File: src/vscode-workspace-extension/src/extension.ts [hunk 4]
```diff
@@ -39,84 +51,181 @@ export function activate(context: vscode.ExtensionContext): void {
...
+        installUserCert(material);
+      }
+
+      for (const dest of parsed.destinations) {
+        const result = writeExtraDestination(dest, material);
+        for (const err of result.errors) log(err);
+        if (result.rehashDir) rehashDirs.add(result.rehashDir);
+      }
+      successes++;
+    } catch (err: unknown) {
```
[the remark this answers is not in the document]
Comment (dnegstad) [reply src/vscode-workspace-extension/src/extension.ts:114]: Updated to track successful vs. actually written certificates and report them separately.

File: README.md [hunk 5]
```diff
@@ -121,6 +121,59 @@ test/
...
+
+| Format | Writes |
+|--------|--------|
+| `pem` | `{name}.pem` (cert only) |
+| `key` | `{name}.key` (private key; skipped when no key is available) |
+| `pem-bundle` | `{name}-bundle.pem` (cert + key concatenated) |
+| `pfx` | `{name}.pfx` (skipped when no key is available) |
+| `all` *(default)* | all of the above |
+
+Directory targets (trailing `/`) rehash the directory once per write so OpenSSL can discover the PEMs.
```
[the remark this answers is not in the document]
Comment (dnegstad) [reply README.md]: Reworded hashing behavior in the docs.

File: src/vscode-ui-extension/src/certProvider.ts [hunk 6]
```diff
@@ -59,26 +158,105 @@ export class CertProvider {
...
   }
 
-  clearCache(): void {
-    this.cached = null;
+  private async loadUserCert(
+    config: UserCertificateConfig
+  ): Promise<CertMaterialV2 | null> {
+    if (!config.name) {
+      throw new Error("userCertificates entry is missing a 'name'.");
+    }
```
[the remark this answers is not in the document]
Comment (dnegstad) [reply src/vscode-ui-extension/src/certProvider.ts]: Added guards against path separators in both the runtime configuration handling and the package.json property definition.

File: src/devcontainer-feature/src/devcontainer-dev-certs/install.sh [hunk 7]
```diff
@@ -49,13 +93,26 @@ if [ "${SSL_CERT_DIRS}" != "${DEFAULT_SSL_CERT_DIRS}" ]; then
     echo "Overriding SSL_CERT_DIR: ${SSL_CERT_DIR_VALUE}"
 fi
 
+# Surface the feature options to the running container so the remote extension
+# can read them via process.env.
+{
+    echo "DEVCONTAINER_DEV_CERTS_GENERATE_DOTNET=${GENERATE_DOTNET_CERT}"
+    echo "DEVCONTAINER_DEV_CERTS_SYNC_USER=${SYNC_USER_CERTIFICATES}"
+    echo "DEVCONTAINER_DEV_CERTS_EXTRA_DESTINATIONS=${EXTRA_CERT_DESTINATIONS}"
+} >> /etc/environment
```
[the remark this answers is not in the document]
Comment (dnegstad) [reply src/devcontainer-feature/src/devcontainer-dev-certs/install.sh]: Updated to handle quotes

File: src/devcontainer-feature/src/devcontainer-dev-certs/scripts/setup-cert.sh [hunk 8]
```diff
@@ -2,22 +2,260 @@
...
+            [ "${dest_count}" -eq 0 ] && break
+            dest_path=$(jq -r ".extraDestinations[${j}].path" "${BUNDLE}")
+            dest_format=$(jq -r ".extraDestinations[${j}].format // \"all\"" "${BUNDLE}")
+            write_extra_destination "${dest_path}" "${dest_format}" "${name}" \
+                "${pem_path}" "${pem_key_path}" "${pfx_path}"
+            if [[ "${dest_path}" == */ ]] && { [ "${dest_format}" = "pem" ] || [ "${dest_format}" = "all" ]; }; then
+                REHASH_DIRS+=("${dest_path%/}")
+            fi
+        done
+    done
```
[the remark this answers is not in the document]
Comment (dnegstad) [reply src/devcontainer-feature/src/devcontainer-dev-certs/scripts/setup-cert.sh]: Updated to add extra defensive
